shark-graph/src/test/java/shark/LongScatterSetTest.kt

```diff
@@ -1,142 +1,137 @@
+    assertThat(set)
+        .doesNotContain(TEST_VALUE)
+        .isEmpty()
+  }
 
-      // Value is removed -> size should decrease, other elements should still be in set
-      assertThat(set)
-          .doesNotContain(value)
-          .hasSize(testValues.size - index - 1)
-      for (i in index + 1 until testValues.size) {
-        assertThat(set).contains(testValues[i])
-      }
-    }
+  @Test fun `removing elements with matching hash`() {
+    val set = LongScatterSet()
+    set.add(SAME_MIX_PHI_1)
+    set.add(SAME_MIX_PHI_2)
 
-    // Verify removing same element twice works correctly.
-    set.add(42)
-    set.remove(42)
-    set.remove(42)
+    set.remove(SAME_MIX_PHI_2)
+    set.remove(SAME_MIX_PHI_1)
 
     assertThat(set)
+        .doesNotContain(SAME_MIX_PHI_1, SAME_MIX_PHI_2)
         .isEmpty()
-        .doesNotContain(42)
+  }
+
+  @Test fun `release operation cleans set`() {
+    val set = LongScatterSet()
+    set.add(TEST_VALUE)
 
-    // Verify removing elements with matching hash. Remove in reverse order than inserting
-    set += 11
-    set += 14723950898
+    set.release()
 
-    set.remove(14723950898)
-    set.remove(11)
     assertThat(set)
+        .doesNotContain(TEST_VALUE)
         .isEmpty()
-        .doesNotContain(11, 14723950898)
   }
 
-  private fun runAssertionsForAddOperation(set: LongScatterSet) {
-    require(set.size() == 0) { "Set should be empty in order to run the assertions on it" }
+  @Test fun `setting initial capacity before operations`() {
+    val set = LongScatterSet()
+    set.ensureCapacity(TEST_CAPACITY)
 
-    // First, verify adding unique values, as well as values that have matching hashKeys
-    // Values 11 and 14723950898 have same hash when it's calculated via HHPC.mixPhi()
-    Assertions.assertThat(HHPC.mixPhi(14723950898)).isEqualTo(HHPC.mixPhi(11))
+    set.add(TEST_VALUE)
+    set.remove(TEST_VALUE)
 
-    val testValues = listOf(42, 0, Long.MIN_VALUE, Long.MAX_VALUE, -1, 11, 14723950898)
+    assertThat(set).isEmpty()
+  }
 
-    // Verify that adding elements to set actually adds them
-    testValues.forEachIndexed { index: Int, value: Long ->
-      // Verify that value is not yet in the set
-      assertThat(set)
-          .doesNotContain(value)
-          .hasSize(index)
+  @Test fun `setting initial capacity after operations`() {
+    val set = LongScatterSet()
+    set.add(TEST_VALUE)
 
-      // Add element to the set
-      set.add(value)
+    set.ensureCapacity(TEST_CAPACITY)
 
-      // Size should increase by one, element and all previous elements should still be in the set
-      assertThat(set).hasSize(index + 1)
-      for (i in 0 until index + 1) {
-        assertThat(set).contains(testValues[i])
-      }
-    }
+    assertThat(set)
+        .contains(TEST_VALUE)
+        .hasSize(1)
+  }
 
-    // Verify that += operator works exactly like add()
-    set += 30
-    assertThat(set).contains(30)
+  @Test fun `adding a lot of elements causes resizing`() {
```

Comment:
What's "a lot" ? How does the test demonstrate that resizing occurred? Also, resizing is not an interesting part the API, it's an impl detail, what should we test instead?

Reply:
this is a whitebox test - it makes sure that resizing doesn't damage the data. That's implementation details, you're right, but i want to test this scenario as well.

I can recycle it into "Stress test" and test set with million (billion?) addition operations